### PR TITLE
gh-115937: Remove implementation details from inspect.signature() docs

### LINKED ISSUE #115937: Incorrect description of the __signature__ attribute in docs
# Bug report

### Bug description:

Since #106311 its documented in the ``inspect.signature()`` sphinx docs as:
"If the passed object has a ``__signature__`` attribute, this function returns it without further computations."

That looks correct per PEP 362, see the [implementation section](https://peps.python.org/pep-0362/#implementation).  But actual code logic goes far away from this simple description:
https://github.com/python/cpython/blob/8e8ab75d97f51c2850eb8cd711010662d5f1d360/Lib/inspect.py#L2549-L2562

Either we should document this properly (then this is a documentation issue).  Or we could change the code and remove undocumented logic.  As this attribute was documented recently, I'll suggest the second option.

### CPython versions tested on:

CPython main branch

### Operating systems tested on:

_No response_

### Linked PRs
* gh-115984
* gh-116086
* gh-116106
* gh-116107

### PR DESCRIPTION
**Pull request by erlend-aasland**
Co-authored-by: Carol Willing <carolcode@willingconsulting.com>
Co-authored-by: Gregory P. Smith <greg@krypto.org>
Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

* Issue: gh-115937

----
📚 Documentation preview 📚: https://cpython-previews--116086.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/inspect.rst [hunk 1]
```diff
@@ -702,6 +699,12 @@ function.
       Python.  For example, in CPython, some built-in functions defined in
       C provide no metadata about their arguments.
 
+   .. note::
+
+      We may use the :attr:`!__signature__` attribute to create the signature.
```
Comment (JelleZijlstra) [inline Doc/library/inspect.rst]:
```suggestion
      If the passed object has a :attr:`!__signature__` attribute, we may use it to create the signature.
```
A little more concrete (your text makes me think "attribute of what?").

Comment (erlend-aasland) [reply Doc/library/inspect.rst]: I agree, that's better.

**Comment by Gouvernathor** [comment]
I disapprove of this way of solving the issue. @ethanfurman did not explain the rationale for changing the inspect module this way for the sole benefit of enum where it would have been just as easy not to (as #115984 shows).

File: Doc/library/inspect.rst [hunk 2]
```diff
@@ -702,6 +699,13 @@ function.
       Python.  For example, in CPython, some built-in functions defined in
       C provide no metadata about their arguments.
 
+   .. note::
```
Comment (erlend-aasland) [inline Doc/library/inspect.rst]: Perhaps this is more fitting.

```suggestion
   .. impl-detail::
```

Comment (willingc) [reply Doc/library/inspect.rst]: I like making this distinction @erlend-aasland.

**Review by gpshead — APPROVED** [review]
I'd leave the "Consult the source code for current semantics." bit off as that feels implied to be when calling something "an implementation detail". But no big deal either way as it is true.

I do understand the Python user desire for officially describing and documenting `__signature__` behavior as a public API people can rely on.  That's outside the scope of this specific issue as this doc update is just trying to set expectations anywhere the existing implementation detail is mentioned.

**Review by willingc — APPROVED** [review]
@erlend-aasland I think this documentation change with the implementation detail directive accurately reflects the current status of the underlying source code.

**Comment by willingc** [comment]
> I do understand the Python user desire for officially describing and documenting __signature__ behavior as a public API people can rely on. That's outside the scope of this specific issue as this doc update is just trying to set expectations anywhere the existing implementation detail is mentioned.

@Gouvernathor As @gpshead mentions, I think having a separate (from documentation) discussion about behavior as a public API makes good sense. Also, I see you maintain ren'py, thanks for doing so. For years, my daughter used it to create interactive game scenarios (she is now a game developer).

**Comment by skirpichev** [comment]
Should I close #115984 in favor of this approach?

**Comment by erlend-aasland** [comment]
> Should I close #115984 in favor of this approach?

Yes, please.

**Comment by erlend-aasland** [comment]
Thanks for the reviews!

**Comment by Gouvernathor** [comment]
> I think having a separate (from documentation) discussion about behavior as a public API makes good sense.

Following this, I think I will open a non-doc issue about reverting what #100039 changed about signatures.

But aside from that, and contrary to my first impression, that Enum PR didn't actually introduce an incompatible change. The `__signature__` field accepted a `Signature|None` object since around 2014 (raising a TypeError for other types) and now accepts `Signature|str|Callable[[], Signature]|None` which is wider.
While I believe both of these additions are a bad design choice for reasons I explained in the linked issue, I think it would still be good to document that passing a Signature object makes it be returned by inspect.signature, that it's _not_ an implementation detail (contrary to how this pull request presents it), and that any other value aside from None is undocumented and implementation-detail territory.
Such a formulation would keep the cachie/lie behavior I defended above and introduced in the original PEP to remain a documented feature.
I'm not sure how people intervening in this merge request would stand about these two propositions, so I would like feedback on that.

---

> Also, I see you maintain ren'py, thanks for doing so.

You're very welcome. All the best to your daughter.

**Pull request merged**